### PR TITLE
Refactor typing/env to separate the filesystem-related logic

### PR DESCRIPTION
**Pull request by gasche**
This PR, which builds on top of #2227, is again only a refactoring PR: it should not affect the semantics of the compiler (there are two small changes in behavior that I mention below; the rest should be strictly identical, only moving code and data around).

The PR came as I was trying to understand the code of Env that is related to compilation units. `typing/env.ml` has more than 2K lines, and about 300 of them are related to compilation units (reading and writing `cmi` files, consistency checking, various caches, etc.), but they are blended in with the rest in a way that makes it difficult (for me) to follow the logic of this part of the code.

The present PR separates this persistent-files logic to a separate module, Persistent_env, that Env then uses internally. The last commit does the actual file split, all other preceding commits are small changes that were necessary for it, either because they abstract away certain details of the Env module that are affected by the split, or because they disentangle the mess of dependencies within the module that makes the split difficult -- it took me 3 attempts to discover everything that could create a cyclic dependency.

The change might be useful to the Merlin people (cc @trefis, @let-def), as Merlin handles cmi files in a different way from the upstream type-checker.

Now on the two small behavior changes:

1. The `avoid external uses of {add_import,crc_units}` commit merges together three loops that are almost identical, but have a difference in the way `Env.add_import` is called or not on transitive dependencies that have no CRC. This specific change should be reviewed by @lpw25. (Understanding the mess of direct/indirect and weak/strong and transparent/opaque dependencies is my next goal.) Making this merge only makes the overall code cleaner, it's easy to separate into two functions if it turns out to be incorrect.

2. The `remove the current unit name from some exceptions` commit removes some information from Persistent_env exception that was constructed, at raising sites, by using internal state from Env (the name of the current compilation unit). This changes the error messages slightly (see the commit message for an example), but I believe the change is harmless. On the other hand, trying to preserve the previous behavior exactly would be doable, but make the code more complex (I could register the exception-handler in `Env` instead of `Persistent_env`, to have the information available at printing-time, but the irregular structure would be a bit vexing).

### CODE REVIEW

**Comment by gasche** [comment]
I just handled a somewhat-painful rebase on top of #2041 (now merged in trunk). Nothing much to declare (I wasn't sure whether `persistent_structures_of_dir` should stay in Env or move in Persistent_env, I left it in Env for now). @diml, I think you might be interested in (reviewing?) the present PR: I think that some of the code structure changes I made reasonate with some of the parts you touched; maybe the split in this PR would have made your work easier.

**Comment by gasche** [comment]
P.S.: I made sure that the PR is split in several clean commits, and it should be easiest to review commit-by-commit.

**Comment by ghost** [comment]
@gasche sure, I'm having a look

File: typing/env.ml
Comment (ghost) [inline typing/env.ml]: If you put the reference and these functions inside a `Current_unit` module with an explicit signature, you are sure that direct references to `current_unit` won't be reintroduced by accident in the future.

File: typing/persistent_env.ml
Comment (ghost) [inline typing/persistent_env.ml]: Maybe we should use a specific datatype here. I didn't immediately remember what `None` meant. For instance:

```ocaml
type 'a pers_struct_info =
  | Not_found
  | Found of { pers_struct : pers_struct; data : 'a }

type 'a t = {
  persistent_structures : (string, 'a pers_struct_info) Hashtbl.t;
...
}
```

**Review by ghost — COMMENTED** [review]
The changes look correct to me, I just left a couple of suggestions. I don't know either whether the changes related to `add_import` are correct or not.

**Comment by gasche** [comment]
@diml Thanks! I'll adress your comments shortly. Do you have an opnion on the interest of the refactoring -- do you agree that it makes the final code nicer, or does it seems roughly equivalent to you?

**Comment by gasche** [comment]
(I pushed two additional commits to implement the suggestions of @diml)

**Comment by gasche** [comment]
I forgot to say: I grepped the code of `utop` to see how `Env` is used there (in particular I was worried about uses of `add_import`, etc.), and I believe that it would not require any code change with this PR. (Otherwise I would change the PR, I guess.)

File: typing/env.ml
Comment (gasche) [reply typing/env.ml]: Done. (It is not super-important to protect this reference, so we could revert this change. Initially I thought of moving the reference into Persistent_env, which is why I needed this preparatory work.)

File: typing/persistent_env.ml
Comment (gasche) [reply typing/persistent_env.ml]: Indeed, thanks! I used `Found` and `Missing` to avoid ambiguities with the `Not_found` exception (and also `missing` is already used in the API), and unnamed parameters instead of a record because I don't know of a very good name for `data`.

**Comment by ghost** [comment]
I agree that it makes the final code look nicer. I find `Env` a bit hard to digest, so I welcome splitting it into smaller individual parts.

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
BTW, in utop we use cppo to be compatible with several versions of the compiler, so we can adapt to breaking changes in the compiler API.

**Comment by gasche** [comment]
@diml (and I guess @trefis?): merging this now could create rebase pains for #2235, so I propose to delay a bit here until you know what to do there.

**Comment by trefis** [comment]
I haven't looked at the code in this PR in details yet but:
- I don't think this should be too painful for #2235, but it might be for #2231 (which I think should also make it into 4.08)
- I'm surprised by the copyright header on `persistent_env.ml[i]`. Shouldn't the name of the original authors of the code also appear on there? (or, simpler: shouldn't it at least include the names that were present at the top of `env.ml[i]`?)

**Comment by gasche** [comment]
Good point on the copyright; in the past I've just put my copyright when creating new files, but they were completely new.

Re. #2231: given that the present PR is non-urgent, I guess that the most reasonable thing is for me to wait until those recent PRs are settled to merge -- and deal with the rebase pains on my end. Grmbl...

File: typing/env.ml
Comment (trefis) [inline typing/env.ml]: why not `Current_unit_name` then?

File: toplevel/topdirs.ml
Comment (trefis) [inline toplevel/topdirs.ml:124]: Why not put the `try .. with` in `Env.import_crcs` itself? Given that it's the one calling `Consistbl.check`.

Also, it looks like `Toploop` calls that function during initialization, without any `try ... with`; apparently it can't fail, but if it ever did, it would be with an uncaught exception, which doesn't seem very nice.

File: utils/misc.ml
Comment (trefis) [inline utils/misc.ml:892]: Do we really want to grow misc? Couldn't `EnvLazy` be in its own file?

**Comment by gasche** [comment]
I just rebased on top of trunk. I'll re-read the rebased diffs, and my plan is to merge once the CI passes.

File: typing/env.ml
Comment (gasche) [reply typing/env.ml]: done

File: toplevel/topdirs.ml
Comment (gasche) [reply toplevel/topdirs.ml:124]: I prefer for the refactoring commit to stay as close to the existing behavior as possible, and I don't know what Toploop is doing with this function, so I did not change anything here.

File: utils/misc.ml
Comment (gasche) [reply utils/misc.ml:892]: I would rather grow misc than risking to introduce module name conflicts. We can always revisit this later.

**Comment by lpw25** [comment]
Obviously very late to this, but I confirm that the CRC related change is correct.

**Pull request merged**
